crates/optimism/flashblocks/src/service.rs

```diff
@@ -203,13 +205,31 @@ where
         {
             let fut = this.canon_receiver.recv();
             pin!(fut);
-            if fut.poll_unpin(cx).is_ready() {
-                // if we have a new canonical message, we know the currently tracked flashblock is
-                // invalidated
-                if let Some(current) = this.current.take() {
-                    trace!(parent_hash=%current.block().parent_hash(), block_number=current.block().number(), "Clearing current flashblock on new canonical block");
-                    return Poll::Ready(Some(Ok(None)))
-                }
+
+            if let Poll::Ready(Ok(Some(current))) = fut.poll_unpin(cx).map(|state| {
+                state.inspect_err(|e| tracing::error!("{e}")).map(|state| {
+                    this.current
+                        .as_ref()
+                        .map(PendingBlock::parent_hash)
+                        .zip(state.tip_checked().map(RecoveredBlock::hash))
+                        .and_then(
+                            |(parent, latest)| {
+                                if parent == latest {
+                                    None
+                                } else {
+                                    this.current.take()
+                                }
+                            },
+                        )
+                })
+            }) {
```

Comment:
I find this a bit hard to read, perhaps we extract this this into a function that returns `Option<PendingBlock>` then I think we can could replace some of this with `?`

Reply:
@mattsse updated